### PR TITLE
feat: validate predicates given to Read Buffer

### PR DESCRIPTION
**Pull request by e-dard**
Tangentially related to issue #1603 

This PR adds a predicate validation step to query execution operations on the Read Buffer.

## Rationale

When a caller provides a predicate expression that cannot be reasonably evaluated then an error should be returned to the caller indicating the operation could not be executed.

Currently the Read Buffer will just blow up.

An example would be: `"region" > 222.2"` where `region` is a string column.

## Notes

This PR does not fix the issue #1603 of comparing a floating point value against an integer column (or vice versa), though I do have some ideas about how I want to add support for comparing floats/intergers in a future PR.

What it does do is ensure that you will get an error if you compare strings -> numbers, bools -> strings etc. I call out the compatability list in the code below.

### CODE REVIEW

File: read_buffer/src/table.rs [hunk 1]
```diff
@@ -568,6 +581,47 @@ impl MetaData {
...
     }
 
+    /// Determine, based on the table meta data, whether all of the provided
+    /// expressions can be applied. An error is returned if any of the
+    /// expressions cannot be applied to the table data.
+    pub fn validate_exprs<'a>(&self, iter: impl IntoIterator<Item = &'a BinaryExpr>) -> Result<()> {
+        iter.into_iter()
+            .try_for_each(|expr| match self.columns.get(expr.column()) {
+                Some(col_meta) => match (col_meta.logical_data_type, expr.literal()) {
+                    (LogicalDataType::Integer, Literal::Integer(_))
```
Comment (e-dard) [inline read_buffer/src/table.rs:591]: The set of types that we will let literals compare themselves against.

File: read_buffer/src/table.rs [hunk 2]
```diff
@@ -1037,6 +1093,80 @@ mod test {
...
+            vec![BinaryExpr::from(("f64_col", "=", "hello"))],
+            vec![BinaryExpr::from(("f64_col", "=", false))],
+            vec![BinaryExpr::from(("str_col", "=", 10_i64))],
+            vec![BinaryExpr::from(("bool_col", "=", "true"))],
+            // mixture valid/invalid
+            vec![
+                BinaryExpr::from(("time", "=", 100_u64)),
+                BinaryExpr::from(("i64_col", "=", "not good")),
+            ],
+        ];
```
Comment (NGA-TRAN) [inline read_buffer/src/table.rs:1162]: Nice set of test options

File: read_buffer/src/chunk.rs [hunk 3]
```diff
@@ -1314,6 +1318,14 @@ mod test {
...
         // No more data
         assert!(itr.next().is_none());
+
+        // invalid predicate
+        let predicate =
+            Predicate::with_time_range(&[BinaryExpr::from(("env", "=", 22.3))], 100, 205); // filter on time
+
+        assert!(chunk
+            .read_filter("Coolverine", predicate, Selection::All)
+            .is_err());
```
Comment (NGA-TRAN) [inline read_buffer/src/chunk.rs:1328]: Nice

**Review by NGA-TRAN — APPROVED** [review]
LGTM

**Review by alamb — APPROVED** [review]
I wonder if:
1. This PR validates the predicate on all public entrypoints of the ReadBuffer
2. If we will remember to add this validation if/when we add a new entrypoint to the ReadBuffer

For example, what about `satisfies_predicate(&self, predicate: &Predicate) -> bool ` -- are we sure this doesn't need a call too?

https://github.com/influxdata/influxdb_iox/pull/1612/files#diff-a618fb754a20f91b2603d6eb4a572f95e328fecd6d86271ca4df55466175e3ccR519

I wonder if calling `validate_exprs` closer to where the `Predicate` is constructed (e.g. near where `to_read_buffer_predicate` is called) would be easier to reason about coverage and less likely to be forgotten if new entrypoints are added.

For example, somewhere near: https://github.com/influxdata/influxdb_iox/blob/er%2Frefactor%2Fread_buffer%2Fpredicate_error/server/src/db/pred.rs#L19

That being said, I think this PR is a definite improvement as is 👍 Thank you

**Comment by alamb** [comment]
FWIW : I ran the test from https://github.com/influxdata/influxdb_iox/issues/1603 cherry-picked on to the code from this PR and it still seems to generate no results: 

```
git checkout er/refactor/read_buffer/predicate_error
git cherry-pick 1d81300
cargo test -p server
```

Output: 
```

---- db::tests::read_with_predicates stdout ----
thread 'db::tests::read_with_predicates' panicked at 'assertion failed: `(left == right)`
  left: `["+-----+-------------------------------+", "| bar | time                          |", "+-----+-------------------------------+", "| 1   | 1970-01-01 00:00:00.000000010 |", "| 2   | 1970-01-01 00:00:00.000000020 |", "+-----+-------------------------------+"]`,
 right: `["++", "++"]`: 

expected:

[
    "+-----+-------------------------------+",
    "| bar | time                          |",
    "+-----+-------------------------------+",
    "| 1   | 1970-01-01 00:00:00.000000010 |",
    "| 2   | 1970-01-01 00:00:00.000000020 |",
    "+-----+-------------------------------+",
]
actual:

[
    "++",
    "++",
]

', server/src/db.rs:1757:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by e-dard** [comment]
> FWIW : I ran the test from #1603 cherry-picked on to the code from this PR and it still seems to generate no results:
> 
> ```
> git checkout er/refactor/read_buffer/predicate_error
> git cherry-pick 1d81300
> cargo test -p server
> ```
> 
> Output:
> 
> ```
> 
> ---- db::tests::read_with_predicates stdout ----
> thread 'db::tests::read_with_predicates' panicked at 'assertion failed: `(left == right)`
>   left: `["+-----+-------------------------------+", "| bar | time                          |", "+-----+-------------------------------+", "| 1   | 1970-01-01 00:00:00.000000010 |", "| 2   | 1970-01-01 00:00:00.000000020 |", "+-----+-------------------------------+"]`,
>  right: `["++", "++"]`: 
> 
> expected:
> 
> [
>     "+-----+-------------------------------+",
>     "| bar | time                          |",
>     "+-----+-------------------------------+",
>     "| 1   | 1970-01-01 00:00:00.000000010 |",
>     "| 2   | 1970-01-01 00:00:00.000000020 |",
>     "+-----+-------------------------------+",
> ]
> actual:
> 
> [
>     "++",
>     "++",
> ]
> 
> ', server/src/db.rs:1757:9
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```

I did mention at the very top of the PR that this isn't a fix for #1603.

**Comment by e-dard** [comment]
> I wonder if:
> 
> 1. This PR validates the predicate on all public entrypoints of the ReadBuffer

In order to operate on any data you need to get row groups that meet the predicate. In order to get row groups you need to run the predicate validation code.

> 2. If we will remember to add this validation if/when we add a new entrypoint to the ReadBuffer
> 
> For example, what about `satisfies_predicate(&self, predicate: &Predicate) -> bool ` -- are we sure this doesn't need a call too?

Yep, that's the one example I forgot (because it doesn't use the normal code-path). I will add validation there before merging.


> I wonder if calling `validate_exprs` closer to where the `Predicate` is constructed (e.g. near where `to_read_buffer_predicate` is called) would be easier to reason about coverage and less likely to be forgotten if new entrypoints are added.

Yeah I thought about that and I think it's a good idea, but it will be simpler once I have got the "one-table-per-chunk" branch I have ongoing merged in, because then there will be a single schema per chunk. I'll look at exposing it as a follow on. Files https://github.com/influxdata/influxdb_iox/issues/1614

**Comment by carols10cents** [comment]
> In order to operate on any data you need to get row groups that meet the predicate. In order to get row groups you need to run the predicate validation code.

If this is true, then this:

> Yep, that's the one example I forgot (because it doesn't use the normal code-path). I will add validation there before merging.

should have caused a compiler error... what about a little typestate, like `validate_exprs` returns a `ValidatedMetaData` and the places that should only use `MetaData` after `validate_exprs` is called would take `ValidatedMetaData` instead of `MetaData`?

**Comment by alamb** [comment]
Closing this PR as it seems to have become stale. Please feel free to reopen

**Pull request closed without merging**